src/core/types.rs

```diff
@@ -0,0 +1,303 @@
+                debug!("Found shorter depth to '{}' new depth is {}", on_node.name, current_depth);
+                on_node.depth = Some(current_depth);
+            }
+        } else {
+            debug!("Initializing depth of node '{}' to {}", on_node.name, current_depth);
+            // If uninitialized, initialize 
+            on_node.depth = Some(current_depth);
+        }
+    }
+
+    pub fn add(&mut self, node: DepNode) -> Ref<DepNode> {
+        assert!(!self.nodes.contains_key(&node.name));
+        debug!("Adding node to graph: {}", node.name);
+
+        let name = node.name.clone(); // TODO: Better way?
+
+        self.nodes.insert(
+            name.clone(),
+            RefCell::new(node)
+        );
+        self.root_nodes.insert(name.clone());
+
+        self.nodes.get(&name).unwrap().borrow()
+    }
+
+    pub fn add_graph(&mut self, from: &str, on: &str, graph: DepGraph) {
+        debug!(
+            "Adding graph with {} nodes linked by {} -> {}",
+            graph.nodes.len(),
+            from,
+            on
+        );
+        assert!(
+            self.nodes.contains_key(from),
+            "Node does not exist on graph: {}", from
+        );
+        assert!(
+            graph.nodes.contains_key(on),
+            "Node does not exist on graph: {}", on
+        );
+
+        // Uninitialize depth from previous graph
+        for (_, node_cell) in &graph.nodes {
+            let mut node = node_cell.borrow_mut();
+            node.depth = None;
+        }
+
+        // TODO: Really should be a better way than cloning here
+        let from_node = self.nodes.get(from).unwrap().borrow().clone();
+        let on_node = graph.nodes.get(on).unwrap().borrow().clone();
+
+        // Construct the depth from the 
+        let mut current_depth = from_node.depth.unwrap();
+        let mut to_process = vec![on_node.name.clone()];
+        while let Some(name) = to_process.pop() {
+            let mut node = graph.nodes.get(&name).unwrap().borrow_mut();
+
+            // If node depth is already set it has been set by a previous path
+            if node.depth.is_none() {
+                node.depth = Some(current_depth);
+
+                // Make sure we process the dependencies of this node
+                to_process.extend(node.dependencies.iter().cloned());
+                current_depth += 1;
+            }
+        }
+
+        // Finally merge graphs and add dependency on proper nodes
+        self.merge(graph);
+        self.add_dependency(from, on); // On will now be a part of the graph
+    }
+
+    // TODO: Read up on the `where` syntax
+    pub fn with<F>(self, name: &str, f: F) where F: Fn(&DepNode) {
+        let node = self.nodes.get(name).unwrap().borrow();
+        // TODO: Not sure I understand the deref part
+        f(node.deref())
+    }
+
+    pub fn has_node(&self, name: &str) -> bool {
+        self.nodes.contains_key(name)
+    }
+
+    pub fn clone_from(&self, name: &str) -> DepGraph {
```

Comment:
This function looks like it could be pretty heavy with all the clones, and seems like it would be called quite a bit? Having the `DepGraph` hold references to nodes in the cache could speed cache usage up quite a bit. Profiling would show how significant time spent cloning in here actually is